main/src/ca/uwaterloo/flix/language/phase/constraintgeneration/TypingContext.scala

```diff
@@ -0,0 +1,209 @@
+    */
+  private var renv: RigidityEnv = RigidityEnv.empty
+
+  /**
+    * The current level. Incremented and decremented as we enter and exit regions.
+    */
+  private var level: Level = Level.Top
+
+  /**
+    * The typing context from outside the current scope.
+    * We push and pop information from this stack when we enter and exit regions.
+    */
+  private val nest: mutable.Stack[ScopeInfo] = mutable.Stack.empty
+
+  /**
+    * Returns the current rigidity environment.
+    */
+  def getRigidityEnv: RigidityEnv = renv
+
+  /**
+    * Returns the current typing constraints.
+    */
+  def getTypingConstraints: List[TypingConstraint] = currentScopeInfo.constrs.toList
+
+  /**
+    * Generates constraints unifying the given types.
+    */
+  def unifyTypeM(tpe1: Type, tpe2: Type, loc: SourceLocation): Unit = {
+    currentScopeInfo.constrs.append(TypingConstraint.Equality(tpe1, tpe2, Provenance.Match(tpe1, tpe2, loc)))
+  }
+
+  /**
+    * Generates constraints unifying the given types.
+    */
+  def unifyAllTypesM(tpes: List[Type], kind: Kind, loc: SourceLocation)(implicit level: Level, flix: Flix): Type = {
+    tpes match {
+      case tpe1 :: rest =>
+        rest.foreach(unifyTypeM(tpe1, _, loc))
+        tpe1
+      case Nil => Type.freshVar(kind, loc.asSynthetic)
+    }
+  }
+
+  /**
+    * Generates constraints expecting the given type arguments to unify.
+    */
+  def expectTypeArguments(sym: Symbol, expectedTypes: List[Type], actualTypes: List[Type], actualLocs: List[SourceLocation], loc: SourceLocation)(implicit flix: Flix): Unit = {
+    expectedTypes.zip(actualTypes).zip(actualLocs).zipWithIndex.foreach {
+      case (((expectedType, actualType), loc), index) =>
+        val argNum = index + 1
+        val prov = Provenance.ExpectArgument(expectedType, actualType, sym, argNum, loc)
+        currentScopeInfo.constrs.addOne(TypingConstraint.Equality(expectedType, actualType, prov))
+    }
+  }
+
+  /**
+    * Generates constraints unifying the given types.
+    */
+  def unifyType3M(tpe1: Type, tpe2: Type, tpe3: Type, loc: SourceLocation): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+    unifyTypeM(tpe1, tpe3, loc)
+  }
+
+  /**
+    * Generates constraints unifying the given effects.
+    */
+  // TODO ASSOC-TYPES this should actually do something
+  def unifyEffM(tpe1: Type, tpe2: Type, loc: SourceLocation)(implicit flix: Flix): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+  }
+
+  /**
+    * Generates constraints expecting the given types to unify.
+    */
+  def expectTypeM(expected: Type, actual: Type, loc: SourceLocation): Unit = {
+    currentScopeInfo.constrs.append(TypingConstraint.Equality(expected, actual, Provenance.ExpectType(expected, actual, loc)))
+  }
+
+  /**
+    * Generates constraints expecting the given types to unify, binding them to the bound type.
+    */
+  // TODO ASSOC-TYPES what does this do?
+  def expectTypeBindM(expected: Type, actual: Type, bind: Type, loc: SourceLocation): Unit = {
```

Comment:
Do we remember the purpose?

Reply:
Nope, something magical with substitutions but I don't know.

Reply:
removed